### PR TITLE
WIP: Foundry 9 types

### PR DESCRIPTION
**Pull request by xdy**
* **Please check if the PR fulfills these requirements**
- [ ] We use semantic versioning (https://github.com/semantic-release/semantic-release to be specific), so follow https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
All kinds, all over the place. Hopefully none that are breaking.

* **What is the current behavior?** (You can also link to an open issue here)
Everything is every type, pretty much.

* **What is the new behavior (if this is a feature change)?**
Stricter foundry v9 typing.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Probably. But, hopefully not.

* **Other information**:

### CODE REVIEW

File: src/module/utils/sheetUtils.ts [hunk 1]
```diff
@@ -189,20 +188,19 @@ export function getKeyByValue(object:{ [x:string]:unknown; }, value:unknown):str
...
   }
 }
 
-export async function getItemDataFromDropData(data:Record<string, any>):Promise<TwodsixItemData> {
+export async function getItemDataFromDropData(data:Record<string, any>) {
   if (data.pack) {
     // compendium
     const pack = game.packs.find((p) => p.collection === data.pack);
-    if (pack.metadata.type !== 'Item') {
+    if (!pack || pack.metadata.entity !== 'Item') {
```
Comment (marvin9257) [inline src/module/utils/sheetUtils.ts:201]: I thought that packs have now migrated to "type" and not "entity" in v9

Comment (xdy) [reply src/module/utils/sheetUtils.ts:201]: Oh, right, I trusted the foundryvtt-types a bit too much there, I guess they haven't gotten around to updating that bit yet.

File: src/module/sheets/AbstractTwodsixActorSheet.ts [hunk 2]
```diff
@@ -101,15 +102,15 @@ export abstract class AbstractTwodsixActorSheet extends ActorSheet {
     switch (type) {
       case "skills":
         if (!game.settings.get('twodsix', 'hideUntrainedSkills')) {
-          itemData.data.value = game.system.template.Item.skills["value"];
+          const skills: dataTwodsix.Skills = <Skills>game.system.template.Item?.skills;
+          itemData.data.value = skills["value"];
```
Comment (marvin9257) [inline src/module/sheets/AbstractTwodsixActorSheet.ts:106]: maybe this should be 
itemData.data.value = skills?.value";

File: src/module/hooks/updateHits.ts [hunk 3]
```diff
@@ -16,20 +17,21 @@ function getCurrentHits(...args: Record<string, any>[]) {
 }
 
 Hooks.on('preUpdateActor', async (actor:TwodsixActor, update:Record<string, any>) => {
-  if (update.data?.characteristics) {
-    update.data.hits = getCurrentHits(actor.data.data.characteristics, update.data.characteristics);
+  if (update.data?.characteristics && actor.type=== 'traveller') {
```
Comment (marvin9257) [inline src/module/hooks/updateHits.ts:20]: As a thought experiment, I've been thinking about whether we need to have NPC, robot and animal actors.  Do we need to check if a traveller? Or maybe better !="ship" ?

File: src/module/hooks/updateHits.ts [hunk 4]
```diff
@@ -16,20 +17,21 @@ function getCurrentHits(...args: Record<string, any>[]) {
...
   if (update.actorData?.data?.characteristics) {
-    const actor = game.actors.get(token.actorId);
-    update.actorData.data.hits = getCurrentHits(
-      // @ts-ignore
-      actor.data.data.characteristics,
-      token.actorData?.data?.characteristics ?? {},
-      update.actorData.data.characteristics
-    );
+    const actor = <TwodsixActor><unknown>game.actors?.get(token.actorId);
+    if (actor && actor.type === 'traveller') {
```
Comment (marvin9257) [inline src/module/hooks/updateHits.ts:29]: Again is the logic === 'traveller' or != 'ship' ?

**Comment by marvin9257** [comment]
Overall, really impressive.  I only understand some of what has been fixed.
A couple of thoughts to consider:

- The migration code seems a bit out of date (e.g. lots of entity vs. type), do we just need to start afresh?
- There are some changes to the settings code?  Will this conflict with @jonepatr recent updates

**Comment by jonepatr** [comment]
`The migration code seems a bit out of date (e.g. lots of entity vs. type), do we just need to start afresh?`

This is fine. We are removing those in the dynamic-imports PR.

`There are some changes to the settings code? Will this conflict with @jonepatr recent updates?`
I think the updated code is already in this PR?

**Comment by xdy** [comment]
> `There are some changes to the settings code? Will this conflict with @jonepatr recent updates?` I think the updated code is already in this PR?
Yep, I rebased before making the PR.

**Comment by xdy** [comment]
> 

> Overall, really impressive. I only understand some of what has been fixed.
Me too. :)

**Comment by xdy** [comment]
As for the traveller vs ship thing, for now I've taken the easy way, just trying to make sure it's at least not the wrong type.
Longer term I think we should do proper subclasses for Traveller, Ship, etc. (And, really, the same for Items.)
The code would become a lot more readable and less complicated if we do that. But it would most likely require migrations as I'm sure we'll find a bunch of bugs in that area. I think I've found some already, but I just papered over them for now. :)

**Comment by marvin9257** [comment]
I made a build of the code a few commits ago to see if it runs.  It did and the few check I did interacting with the actor sheet worked fine.  But the old style of settings (all on one page persists).  Let me try again

**Comment by xdy** [comment]
Hooray! Deepscan is happy!
Lots of things still broken, but, eh, I'll take a win when I can. :)

**Comment by xdy** [comment]
Will do a manual 1.0.0 release once this has built on master.
